File: TASK07/targets/opcode/vm.go
package opcode

import (
	"fmt"
	"io"
	"os"
)

type vm struct {
	memory  []byte
	pointer int
}

type printBuffer struct {
	buffer []byte

	writer    io.Writer
	autoFlush bool
	length    int
}

func (b *printBuffer) Write(bytes ...byte) {
	b.buffer = append(b.buffer, bytes...)

	if b.autoFlush && len(b.buffer) > b.length {
		b.Flush()
	}
}

func (b *printBuffer) Flush() {
	if len(b.buffer) == 0 {
		return
	}

	b.writer.Write(b.buffer)
	b.buffer = []byte{}
}

func Run(oc []int) {
	v := vm{memory: make([]byte, 30000)}
	buffer := printBuffer{
		writer:    os.Stdout,
		autoFlush: true,
		length:    50,
	}
	length := len(oc)
	for i := 0; i < length; i++ {
		switch Opcode(oc[i]) {
		case ChangeValue:
			pointer := v.pointer + oc[i+1]
			v.memory[pointer] += byte(oc[i+2])
			i += 2
		case InputByte:
			i++
			pointer := v.pointer + oc[i]
			fmt.Scanf("%c", &v.memory[pointer])
		case OutputByte:
			i++
			pointer := v.pointer + oc[i]
			buffer.Write(v.memory[pointer])
		case JumpIfZero:
			i++
			v.pointer += oc[i]

			i++
			jump := oc[i]

			if v.memory[v.pointer] == 0 {
				i += jump
			}
		case JumpIfNotZero:
			i++
			v.pointer += oc[i]

			i++
			jump := oc[i]

			if v.memory[v.pointer] != 0 {
				i -= jump
			}
		case SetValue:
			i++
			pointer := v.pointer + oc[i]

			i++
			value := byte(oc[i])

			v.memory[pointer] = value
		default:
			panic(fmt.Sprintf("opcode: run: invalid opcode %x", uint(oc[i])))
		}
	}

	buffer.Flush()
}

File: TASK08/opcode/vm.cpp
#pragma once
#include <iostream>
#include <vector>
#include <sstream>

#include "./opcode.cpp"
using namespace std;


class VM {
public:
    vector<char> memory;
    int pointer;

    VM(vector<char> mem) {
        memory = mem;
        pointer = 0;
    }
};

class PrintBuffer {
public:
    vector<char> buffer;
    bool autoFlush;
    int length;
    stringstream ss_buf;
    int buf_len;
    PrintBuffer(bool flush, int len) {
        autoFlush = flush;
        length = len;
        buf_len = 0;
    }

    void write(vector<char> bytes) {
        string ret = "";
        for(int i=0; i<bytes.size(); i++) {
            ret += bytes[0];
            buf_len++;
        }
        ss_buf << ret;

        if(autoFlush && buf_len > length) {
            flush();
        }
    }

    void flush() {
        if(buf_len == 0) {
            return;
        }

        cout<<ss_buf.str();
        ss_buf.clear();
        buf_len = 0;
    }
};

void Run(vector<int> oc) {
    vector<char> memory(30000);
    VM vm = VM(memory);
    PrintBuffer buffer = PrintBuffer(true, 50);
    
    int len = oc.size();
    for(int i=0; i<len; i++) {
        switch (oc[i]) {
        case ChangeValue:
            {int pointer = vm.pointer + oc[i+1];
            vm.memory[pointer] += oc[i+2];
            i += 2;}
        case InputByte:
            {i++;
            int pointer = vm.pointer + oc[i];
            cin>>vm.memory[pointer];}
        case OutputByte:
            {i++;
            int pointer = vm.pointer + oc[i];
            buffer.write(vector<char>(vm.memory[pointer]));}
        case JumpIfZero:
            {i++;
            vm.pointer += oc[i];

            i++;
            int jump = oc[i];

            if(vm.memory[vm.pointer] == 0) {
                i += jump;
            }}
        case JumpIfNotZero:
            {i++;
            vm.pointer += oc[i];

            i++;
            int jump = oc[i];
            if(vm.memory[vm.pointer] != 0) {
                i -= jump;
            }}
        case SetValue:
            {i++;
            int pointer = vm.pointer + oc[i];

            i++;
            char value = oc[i];

            vm.memory[pointer] = value;}
        default:
            throw "opcode, run: invalid opcode" + oc[i];
        }
    }
    buffer.flush();
}